### PR TITLE
ARROW-4637: [Python] Conditionally import pandas symbols if they are used. Do not require pandas as a test dependency

### PR DESCRIPTION
**Pull request by wesm**
Warning: hold your nose for this one =) I think this can be made cleaner, but I just wanted to triage all the problems and make sure we don't introduce a hard dependency on pandas again

Also resolves ARROW-4794: make pandas an optional dependency

### CODE REVIEW

**Comment by wesm** [comment]
Import times

before

```
In [1]: %time import pyarrow as pa                                                                                                                               
CPU times: user 155 ms, sys: 16.3 ms, total: 171 ms
Wall time: 187 ms
```

after

```
In [1]: %time import pyarrow as pa
CPU times: user 50.8 ms, sys: 3.91 ms, total: 54.7 ms
Wall time: 76 ms
```

**Comment by wesm** [comment]
Hm, really strange that this fails in the manylinux1 build

```
============================= test session starts ==============================
platform linux2 -- Python 2.7.15, pytest-4.3.1, py-1.8.0, pluggy-0.9.0
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/home/travis/build/apache/arrow/python/manylinux1/.hypothesis/examples')
rootdir: /home/travis/build/apache/arrow/python, inifile: setup.cfg
plugins: hypothesis-4.11.5
collecting ... /home/travis/build/apache/arrow/ci/travis_script_manylinux.sh: line 43: 11744 Segmentation fault      (core dumped) pytest --pyargs pyarrow
The command "if [ $ARROW_CI_PYTHON_AFFECTED == "1" ]; then $TRAVIS_BUILD_DIR/ci/travis_script_manylinux.sh; fi" exited with 139.
```

**Comment by xhochy** [comment]
cc @fjetter

**Comment by wesm** [comment]
I see that no benchmarks were added in https://github.com/apache/arrow/commit/957fe15bb3dafadda4d0a6f9359e4f40107440bb#diff-23eeeb4347bdd26bfc6b7ee9a3b755dd. If we're making a performance improvement, we need to get in the habit of adding benchmarks or we are going to be playing "whack-a-mole" with performance regressions. We are working on having some better infrastructure for getting systematic performance feedback on PRs (cc @kszucs)

**Comment by wesm** [comment]
I'll debug the py2.7 manylinux1 segfault tomorrow and get this sorted out. if you could review in the meantime so I can merge when the build passes that would be great

**Comment by wesm** [comment]
The segfault is coming from the tensorflow/pyarrow import order again. Not sure why this patch causes it

**Comment by wesm** [comment]
I'm going to remove tensorflow from the manylinux1 tests; it doesn't seem worthwhile to try to maintain the brittle equilibrium

**Comment by wesm** [comment]
I suspect the reason that this patch caused things to break is because pandas uses pthreads in various ways (and in its dependencies) and so by making the pandas imports lazy, the pthread initialization is no longer happening when `import pyarrow`

**Comment by wesm** [comment]
This definitely does have a performance regression in the case from ARROW-4629

```
import pandas as pd
import pyarrow as pa
import cProfile
ser = pd.Series(range(10000))
df = pd.DataFrame({col: ser.copy(deep=True) for col in range(100)})
# Simulate a real dataset, i.e. force copy of data
df = df.astype({col: str for col in range(50)})
prof = cProfile.Profile()

prof.enable()
# a few times to collect statistics
for _ in range(500):
    pa.Table.from_pandas(df, nthreads=1)
prof.disable()
prof.dump_stats("array_conversion.prof")
```

I'll fix it (will require some cythonization I think) and add an ASV benchmark for this

**Comment by wesm** [comment]
@xhochy if you could review the code in the meantime so I can merge this once I resolve the perf regression that would be awesome

**Comment by wesm** [comment]
OK, on the following benchmark:

```
import pandas as pd
import pyarrow as pa
import cProfile
ser = pd.Series(range(10000))
df = pd.DataFrame({col: ser.copy(deep=True) for col in range(100)})
# Simulate a real dataset, i.e. force copy of data
df = df.astype({col: str for col in range(50)})
prof = cProfile.Profile()

prof.enable()
# a few times to collect statistics
for _ in range(500):
    pa.Table.from_pandas(df, nthreads=1)
prof.disable()
prof.dump_stats("array_conversion.prof")
```

I have

* master branch: 10.6 seconds runtime
* this patch: 10.0 runtime

So this gives about a 5% speedup. This microperformance stuff is very difficult to keep fast (see eg in pandas https://mail.python.org/pipermail/pandas-dev/2016-March/000499.html) , but I'll add this to ASV

**Comment by wesm** [comment]
OK. Modulo fixing the build this is ready to go

**Comment by kszucs** [comment]
@wesm created an issue to run the ASV bencmarks and host the results https://github.com/ursa-labs/ursabot/issues/16 (this shouldn't take much time)

**Comment by xhochy** [comment]
Looks good but CI is still very unhappy ;)

**Comment by wesm** [comment]
@kszucs I would rather invest the energy in having all the data available in a central benchmark database...

**Comment by wesm** [comment]
I'll fix the CI issues and then merge today

File: python/pyarrow/pandas-shim.pxi [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+            return isinstance(obj, self._series)
+        else:
+            return False
+
+    def assert_frame_equal(self, *args, **kwargs):
+        self._check_import()
+        return self._pd.util.testing.assert_frame_equal
+
+
+cdef _PandasAPIShim pandas_api = _PandasAPIShim()
```
Comment (kszucs) [inline python/pyarrow/pandas-shim.pxi]: `cpdef`?

File: python/pyarrow/pandas-shim.pxi [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# pandas lazy-loading API shim that reduces API call and import overhead
+
+cdef class _PandasAPIShim(object):
```
Comment (kszucs) [inline python/pyarrow/pandas-shim.pxi]: I don't like it but get your point :)
We should have a ticket about refactoring this.

Comment (wesm) [reply python/pyarrow/pandas-shim.pxi]: Microperformance is a huge pain. The code that touches this is very sensitive to changes. If you _do_ refactor, you must be extremely paranoid about making things slower

For example I noticed that the operation

```
import pyarrow as pa
import numpy as np

arr = np.array([1, 2, 3, 4])[::2]
arrow_arr = pa.array(arr)
```

can get 10-20% slower with tiny modifications to things. Suppose that you have an operation that takes 3 microseconds, and you make changes that cause it to take 5 microseconds. Then you're dropping 2 seconds for every million calls. 

see https://mail.python.org/pipermail/pandas-dev/2016-March/000499.html

**Comment by wesm** [comment]
I fixed at least the Python 2.7 issue. Will merge as soon as the build is green

**Comment by wesm** [comment]
Looks like there's one remaining issue with ORC Python bindings on Windows. Looking

**Comment by wesm** [comment]
Haha here I'm trying to debug this and I missed the critical thing:

```
if(MSVC)
  # ORC doesn't build on windows
  set(ARROW_ORC OFF)
  # Plasma using glog is not fully tested on windows.
  set(ARROW_USE_GLOG OFF)
endif()
```

Not sure why it's only trying to run the ORC tests now. I'll fix

**Comment by wesm** [comment]
Appyveyor build: https://ci.appveyor.com/project/wesm/arrow/builds/23193451

**Comment by malthe** [comment]
The downside of this fix is that it causes import deadlocking if using Pandas in multithreaded mode.

**Pull request closed without merging**